### PR TITLE
Add presence subscribe with presence action APIs

### PR DESCRIPTION
**Pull request by gokhanbarisaker**
Resolves: #21

### CODE REVIEW

File: lib/src/main/java/io/ably/lib/realtime/Presence.java [hunk 1]
```diff
@@ -13,12 +13,7 @@
 import io.ably.lib.types.ProtocolMessage;
 import io.ably.lib.util.Log;
 
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.Iterator;
-import java.util.Map;
-import java.util.Set;
+import java.util.*;
```
Comment (mattheworiordan) [inline lib/src/main/java/io/ably/lib/realtime/Presence.java]: Wy did you change this?

**Comment by mattheworiordan** [comment]
I am reasonably confident that this has broken the `unsubscribe(listener)` functionality in that it will not remove the `eventListeners` registered for that listener, see https://github.com/gokhanbarisaker/ably-java/blob/presence-subscribe-filter/lib/src/main/java/io/ably/lib/realtime/Presence.java#L82.

Please can you add a test that adds an event based listener, then calls unsubscribe all, and makes sure the listener has been removed.

While you are looking at this code, can you add a method `unsubscribe()` that simply removes all event and all-event listeners?  This is defined in ["the IDL"](http://docs.ably.io/client-lib-development-guide/features/#idl).  `unsubscribe()` is supported for both channels and presence.

**Comment by paddybyers** [comment]
> Please can you add a test that adds an event based listener, then calls unsubscribe all, and makes sure the listener has been removed.

Was this done?

**Comment by gokhanbarisaker** [comment]
@paddybyers Below parts were missing,

  - https://github.com/ably/ably-java/pull/97/files#diff-db9d112187c909d918e250dd2544de78R271
  - https://github.com/ably/ably-java/pull/97/files#diff-fc7f109a620e3e0188e0516eb98ed0f5R90

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java [hunk 2]
```diff
@@ -1288,4 +1283,405 @@ public void disconnect_leave() {
...
+			/* Wait until receiver client (ably2) observes {@code Action.leave}
+			 * is emitted from emitter client (ably1)
+			 */
+			Helpers.PresenceWaiter leavePresenceWaiter = new Helpers.PresenceWaiter(channel2);
+			leavePresenceWaiter.waitFor(ably1.options.clientId, Action.leave);
+
+			/* Validate that,
+			 *	- we received all actions
+			 */
+			assertThat(receivedMessageStack.size(), is(equalTo(3)));
```
Comment (mattheworiordan) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: Shouldn't this be 4, 4 presence events have been emitted in the above code?

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java [hunk 3]
```diff
@@ -1288,4 +1283,405 @@ public void disconnect_leave() {
...
+			Channel channel1 = ably1.channels.get(channelName);
+			channel1.attach();
+			(new ChannelWaiter(channel1)).waitFor(ChannelState.attached);
+
+			Channel channel2 = ably2.channels.get(channelName);
+			channel2.attach();
+			(new ChannelWaiter(channel2)).waitFor(ChannelState.attached);
+
+			/* Start listening channel for events with ably client 2 (receiver) */
+			channel2.presence.enter("Hello, #1", null);
```
Comment (mattheworiordan) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: You don't have to be present to subscribe to presence events, I would recommend removing this enter of presence for channel2 as it's unrelated to the test and we should be sure that presence events are published to clients who are not present

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java [hunk 4]
```diff
@@ -1288,4 +1283,405 @@ public void disconnect_leave() {
...
+			/* Wait until receiver client (ably2) observes {@code Action.leave}
+			 * is emitted from emitter client (ably1)
+			 */
+			Helpers.PresenceWaiter leavePresenceWaiter = new Helpers.PresenceWaiter(channel2);
+			leavePresenceWaiter.waitFor(ably1.options.clientId, Action.leave);
+
+			/* Validate that,
+			 *	- we received specific actions
+			 */
+			assertThat(receivedMessageStack.size(), is(equalTo(3)));
```
Comment (mattheworiordan) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: As above, should this not be 4?

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java [hunk 5]
```diff
@@ -1288,4 +1283,405 @@ public void disconnect_leave() {
...
+			Channel channel1 = ably1.channels.get(channelName);
+			channel1.attach();
+			(new ChannelWaiter(channel1)).waitFor(ChannelState.attached);
+
+			Channel channel2 = ably2.channels.get(channelName);
+			channel2.attach();
+			(new ChannelWaiter(channel2)).waitFor(ChannelState.attached);
+
+			/* Start listening channel for events with ably client 2 (receiver) */
+			channel2.presence.enter("Hello, #1", null);
```
Comment (mattheworiordan) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: As above, this client does not need to enter themselves

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java [hunk 6]
```diff
@@ -1288,4 +1283,405 @@ public void disconnect_leave() {
...
+
+			/* Wait until receiver client (ably2) observes {@code Action.leave}
+			 * is emitted from emitter client (ably1)
+			 */
+			new Helpers.PresenceWaiter(channel2).waitFor(ably1.options.clientId, Action.leave);
+
+			/* Validate that,
+			 *	- we received specific actions
+			 */
+			assertThat(receivedMessageStack, is(not(Matchers.<PresenceMessage>empty())));
```
Comment (mattheworiordan) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: Shouldn't this be one event if the user is subscribing to enter events?

Comment (gokhanbarisaker) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: We are validating,
  - client received more than 0 `PresenceMessage`.
  - all received presence messages have `enter` action.

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java (hunk 5 above)
Comment (gokhanbarisaker) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: done

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java (hunk 3 above)
Comment (gokhanbarisaker) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: done

File: lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java (hunk 2 above)
Comment (gokhanbarisaker) [inline lib/src/test/java/io/ably/lib/test/realtime/RealtimePresenceTest.java]: done

**Comment by gokhanbarisaker** [comment]
@mattheworiordan, I have rebased & applied requested fixes

**Comment by mattheworiordan** [comment]
LGTM, @paddybyers?

**Comment by mattheworiordan** [comment]
@paddybyers?

**Comment by paddybyers** [comment]
LGTM

**Pull request merged**